File: rust-2021/src/day10.rs
type I = usize;
type Input = (I, I);

pub fn read(inp: &str) -> Input {
    let mut v1 = 0;
    let mut v2 = Vec::with_capacity(100);
    for line in inp.trim().split('\n') {
        let (c1, c2) = check(line);
        if c1 != 0 {
            v1 += c1;
        } else {
            v2.push(c2)
        }
    }
    let n = v2.len() >> 1;
    (v1, *v2.select_nth_unstable(n).1)
}

#[inline]
fn check(line: &str) -> (I, I) {
    let mut lifo = Vec::with_capacity(20);
    for c in line.chars() {
        match c {
            '(' | '{' | '[' | '<' => lifo.push(c),
            ')' => {
                if lifo.pop() != Some('(') {
                    return (3, 0);
                }
            }
            ']' => {
                if lifo.pop() != Some('[') {
                    return (57, 0);
                }
            }
            '}' => {
                if lifo.pop() != Some('{') {
                    return (1197, 0);
                }
            }
            '>' => {
                if lifo.pop() != Some('<') {
                    return (25137, 0);
                }
            }
            _ => {
                unreachable!()
            }
        }
    }
    let mut s = 0;
    for &c in lifo.iter().rev() {
        s = s * 5
            + match c {
                '(' => 1,
                '[' => 2,
                '{' => 3,
                '<' => 4,
                _ => unreachable!(),
            }
    }
    (0, s)
}

pub fn part1(inp: &Input) -> I {
    inp.0
}

pub fn part2(inp: &Input) -> I {
    inp.1
}

pub fn run() -> (String, String) {
    let inp = include_str!(concat!(env!("CARGO_MANIFEST_DIR"), "/inputs/day10.input"));
    let data = read(inp);
    let a = part1(&data).to_string();
    let b = part2(&data).to_string();
    (a, b)
}

File: 2022/day10.py
import time

start = time.time()
with open("../inputs/2022/day10.input") as f:
    data = f.read()
    inp = [
        0 if line == "noop" else int(line.split(" ")[1]) for line in data.split("\n")
    ]


def part1(inp: list) -> int:
    X = 1
    result = 0
    cycle = 1
    check = 20
    for op in inp:
        cycle += 1 + (op != 0)
        if cycle > check:
            result += X * check
            check += 40
        X += op
    return result


ALPHABET = {
    " ## \n#  #\n#  #\n####\n#  #\n#  #": "A",
    "### \n#  #\n### \n#  #\n#  #\n### ": "B",
    " ## \n#  #\n#   \n#   \n#  #\n ## ": "C",
    "####\n#   \n### \n#   \n#   \n####": "E",
    "####\n#   \n### \n#   \n#   \n#   ": "F",
    " ## \n#  #\n#   \n# ##\n#  #\n ###": "G",
    "#  #\n#  #\n####\n#  #\n#  #\n#  #": "H",
    " ###\n  # \n  # \n  # \n  # \n ###": "I",
    "  ##\n   #\n   #\n   #\n#  #\n ## ": "J",
    "#  #\n# # \n##  \n# # \n# # \n#  #": "K",
    "#   \n#   \n#   \n#   \n#   \n####": "L",
    " ## \n#  #\n#  #\n#  #\n#  #\n ## ": "O",
    "### \n#  #\n#  #\n### \n#   \n#   ": "P",
    "### \n#  #\n#  #\n### \n# # \n#  #": "R",
    " ###\n#   \n#   \n ## \n   #\n### ": "S",
    "#  #\n#  #\n#  #\n#  #\n#  #\n ## ": "U",
    "#   \n#   \n # #\n  # \n  # \n  # ": "Y",
    "####\n   #\n  # \n #  \n#   \n####": "Z",
}


def part2(inp: list) -> str:
    W = 40
    H = 6

    result: list[str] = [" "] * (W * H)

    x = 1
    cycle = 0
    for op in inp:
        for _ in range(1 + (op != 0)):
            result[cycle] = " " if abs(x - (cycle % W)) > 1 else "#"
            cycle += 1
        x += op

    return "".join(
        ALPHABET[
            "\n".join(
                "".join(result[W * row + c * 5 : W * row + c * 5 + 4])
                for row in range(H)
            )
        ]
        for c in range(8)
    )


print(
    f"Day 10     {part1(inp):<14} {part2(inp):<14} {(time.time() - start)*1e3:>11.2f}"
)
